### PR TITLE
BUG: fix CubicSpline(..., bc_type="periodic") #11758

### LINKED ISSUE #11758: BUG: of scipy.interpolate.CubicSpline when `bc_type' is set to `periodic'
in line 701 of file: https://github.com/scipy/scipy/blob/c1372d8aa90a73d8a52f135529293ff4edb98fc8/scipy/interpolate/_cubic.py#L701

                a_m2_m1 = dx[-2]
should be

                a_m2_m1 = dx[-3]

This typo could make the result of some interpolations quite different.

### PR DESCRIPTION
**Pull request by egorz734**
#### Reference issue

Closes #11758
#### What does this implement/fix?

1. `a_m2_m1 = dx[-2]` was changed to `a_m2_m1 = dx[-3]` because it implements the `c[n-2]` (lower right corner) element of the matrix where `c` is the upper diagonal formed by following equations:
https://github.com/scipy/scipy/blob/ab9e9f17e0b7b2d618c4d4d8402cd4c0c200d6c0/scipy/interpolate/_cubic.py#L681-L686
where `c[i]` is equal to `x[n-3] - x[n-4] = dx[-3]` when `i = n - 3`.
2. In 3 points case all the first derivatives are equal and (also due to the first fix) we changed it with manual calculations.
#### Additional information
Co-authored-by: Diana Sukhoverkhova <diana.suhoverhova@mail.ru>

### CODE REVIEW

File: scipy/interpolate/_cubic.py [hunk 1]
```diff
@@ -562,6 +562,10 @@ class CubicSpline(CubicHermiteSpline):
     and `InterpolatedUnivariateSpline`, but these two methods use a
     representation in B-spline basis.
 
+    When boundary condition is 'periodic' and n = 3, all the first derivatives
+    are equal due to the special form of the matrix ''A'', so this case should
+    be held manually.
```
Comment (ev-br) [inline scipy/interpolate/_cubic.py:567]: This is not needed here.

File: scipy/interpolate/_cubic.py [hunk 2]
```diff
@@ -694,53 +698,60 @@ def __init__(self, x, y, axis=0, bc_type='not-a-knot', extrapolate=None):
...
-                s = np.empty((n,) + y.shape[1:], dtype=y.dtype)
-                s[:-2] = s1 + s_m1 * s2
-                s[-2] = s_m1
-                s[-1] = s[0]
+                # In case when number of points is 3 we should count derivatives
+                # manually
+                if n == 3:
+                    s = np.empty((n,) + y.shape[1:], dtype=y.dtype)
+                    t = (slope / dxr).sum() / (1. / dxr).sum()
+                    s.fill(t)
```
Comment (ev-br) [inline scipy/interpolate/_cubic.py:706]: To minimize the diff size, better move this special case to up, close to `if n === 3 and bc_type[0] == ... `. This way the general case does not get extra indent, and the actual bug fix is visible in the diff.

**Review by ev-br — CHANGES_REQUESTED** [review]
Looks good modulo two minor comments inline.

Full disclosure: Egor and Diana work with me at HSE, so it'd be good to have a review from someone else, too.

File: scipy/interpolate/tests/test_polyint.py [hunk 3]
```diff
@@ -604,6 +604,23 @@ def test_periodic_eval(self):
...
+        S = CubicSpline(x, y, bc_type='periodic', extrapolate='periodic')
+        self.check_correctness(S, 'periodic', 'periodic')
+
+    def test_three_points(self):
+        # gh-11758: Fails computing a_m2_m1
+        x = np.array([1.0, 2.75, 3.0])
+        y = np.array([1.0, 15.0, 1.0])
+        S = CubicSpline(x, y, bc_type='periodic')
+        self.check_correctness(S, 'periodic', 'periodic')
+        assert_allclose(S.derivative(1)(x), np.array([-48.0, -48.0, -48.0]))
```
Comment (ev-br) [inline scipy/interpolate/tests/test_polyint.py:622]: Why this value? Please add a short comment on how to arrive to this number.

**Review by ev-br — APPROVED** [review]
LGTM modulo a small comment.

File: scipy/interpolate/tests/test_polyint.py (hunk 3 above)
Comment (diana-sukhoverkhova) [reply scipy/interpolate/tests/test_polyint.py:622]: In three points case, unknown vector `s` (first derivatives) could be found manually by solving system of 2 linear equations. Due to solution of this system, all `s[i]` are equal and could be computed according to the following formula:



where `h1=x[1]-x[0]`, `h2=x[2]-x[1]`, `m1=y[1]-y[0]/h1`, `m2=y[2]-y[1]/h2`

Comment (ev-br) [reply scipy/interpolate/tests/test_polyint.py:622]: Could add this to explanation as a comment

Comment (rgommers) [reply scipy/interpolate/tests/test_polyint.py:626]: Explanation was added a few lines up.

**Review by rgommers — APPROVED** [review]
This looks good to me, merging. Thanks @egorz734 and @diana-sukhoverkhova.

**Pull request merged**